lib/puppet/file_system/windows.rb

```diff
@@ -89,6 +99,25 @@ def read_preserve_line_endings(path)
 
   private
 
+  def expand_path_cached(path)
+    @memoized_expanded_paths ||= {}
+    split_index = path.index('/', path.index('~'))
+    to_expand = path[0..(split_index -1)]
+    if !@memoized_expanded_paths.key?(to_expand)
+      expanded = ''
+
+      # Ruby limitation: must call expand_path on each subdirectory
```

Comment:
This is mostly the reason this PR exists.  Read the docs for the original `expanded_path_string`